projects/rocprofiler-systems/source/lib/core/gpu_metrics.cpp

```diff
@@ -0,0 +1,330 @@
+{
+    std::vector<uint16_t> values;
+    values.reserve(count);
+    for(uint8_t i = 0; i < count; ++i)
+        values.push_back(deserialize_uint16(data, offset));
+    return values;
+}
+
+std::vector<uint64_t>
+deserialize_uint64_vector(const std::vector<uint8_t>& data, size_t& offset, uint8_t count)
+{
+    std::vector<uint64_t> values;
+    values.reserve(count);
+    for(uint8_t i = 0; i < count; ++i)
+        values.push_back(deserialize_uint64(data, offset));
+    return values;
+}
+}  // namespace
+
+std::vector<uint8_t>
+serialize_gpu_metrics(const gpu_metrics_t& metrics, bool is_vcn_activity_supported,
+                      bool                          is_jpeg_activity_supported,
+                      const gpu_metrics_settings_t& settings)
+{
+    // Flatten XCP data if needed and pre-calculate counts
+    // Example:
+    // XCP 0: [10, 20, 30]        (3 values)
+    // XCP 1: [15, 25]            (2 values)
+    // XCP 2: [5, 10, 15, 20]     (4 values)
+    // vcn_xcp_count: 3
+    // vcn_xcp_sizes: [3, 2, 4]
+    // vcn_data_flat: [10, 20, 30, 15, 25, 5, 10, 15, 20]
+    std::vector<uint16_t> vcn_data_flat;
+    std::vector<uint16_t> jpeg_data_flat;
+    std::vector<uint8_t>  vcn_xcp_sizes;   // Size of each XCP's VCN data
+    std::vector<uint8_t>  jpeg_xcp_sizes;  // Size of each XCP's JPEG data
+
+    if(is_vcn_activity_supported)
+    {
+        vcn_data_flat = metrics.vcn_activity;
+    }
+    else
+    {
+        // Flatten per-XCP VCN data and record sizes
+        for(const auto& xcp_data : metrics.vcn_busy)
+        {
+            vcn_xcp_sizes.push_back(static_cast<uint8_t>(xcp_data.size()));
+            vcn_data_flat.insert(vcn_data_flat.end(), xcp_data.begin(), xcp_data.end());
+        }
+    }
+
+    if(is_jpeg_activity_supported)
+    {
+        jpeg_data_flat = metrics.jpeg_activity;
+    }
+    else
+    {
+        // Flatten per-XCP JPEG data and record sizes
+        for(const auto& xcp_data : metrics.jpeg_busy)
+        {
+            jpeg_xcp_sizes.push_back(static_cast<uint8_t>(xcp_data.size()));
+            jpeg_data_flat.insert(jpeg_data_flat.end(), xcp_data.begin(), xcp_data.end());
+        }
+    }
+
+    uint8_t vcn_count        = static_cast<uint8_t>(vcn_data_flat.size());
+    uint8_t jpeg_count       = static_cast<uint8_t>(jpeg_data_flat.size());
+    uint8_t vcn_xcp_count    = static_cast<uint8_t>(vcn_xcp_sizes.size());
+    uint8_t jpeg_xcp_count   = static_cast<uint8_t>(jpeg_xcp_sizes.size());
+    uint8_t xgmi_read_count  = static_cast<uint8_t>(metrics.xgmi_read_data_acc.size());
+    uint8_t xgmi_write_count = static_cast<uint8_t>(metrics.xgmi_write_data_acc.size());
+
+    std::vector<uint8_t> result;
+
+    // Serialize capability flags
+    uint8_t flags = 0;
+    if(is_vcn_activity_supported) flags |= (1 << 0);  // Set bit for VCN activity support
+    if(is_jpeg_activity_supported)
+        flags |= (1 << 1);  // Set bit for JPEG activity support
+    serialize_uint8(result, flags);
```

Comment:
```suggestion
    // Serialize capability flags
    struct {
        union {
            struct {
                uint8_t vcn_activity : 1;
                uint8_t jpeg_activity : 1;
            };
            uint8_t value;
        };
    } capabilities;

    capabilities.vcn_activity = is_vcn_activity_supported ? 1 : 0;
    capabilities.jpeg_activity = is_jpeg_activity_supported ? 1 : 0;
    serialize_uint8(result, capabilities.value);
```

If you want to make "flag setting" a little more readable, I suggest a structure like above.